src/devices/src/virtio/queue.rs

```diff
@@ -309,31 +330,42 @@ impl Queue {
     }
 
     /// Puts an available descriptor head into the used ring for use by the guest.
-    pub fn add_used(&mut self, mem: &GuestMemoryMmap, desc_index: u16, len: u32) {
+    pub fn add_used(
+        &mut self,
+        mem: &GuestMemoryMmap,
+        desc_index: u16,
+        len: u32,
+    ) -> Result<(), QueueError> {
         if desc_index >= self.actual_size() {
             error!(
                 "attempted to add out of bounds descriptor to used ring: {}",
                 desc_index
             );
-            return;
+            return Ok(());
         }
 
         let used_ring = self.used_ring;
         let next_used = u64::from(self.next_used.0 % self.actual_size());
         let used_elem = used_ring.unchecked_add(4 + next_used * 8);
 
-        // These writes can't fail as we are guaranteed to be within the descriptor ring.
-        mem.write_obj(u32::from(desc_index), used_elem).unwrap();
-        mem.write_obj(len as u32, used_elem.unchecked_add(4))
-            .unwrap();
+        mem.write_obj(u32::from(desc_index), used_elem)
+            .map_err(|_e| QueueError::InvalidUsedRingAddress(desc_index as u32, used_elem))?;
```

Comment:
This error is not necessarily related to an InvalidAddress. It can be any of the [following](https://github.com/rust-vmm/vm-memory/blob/171e91e1bb057d3af9a65adf1e45952f24da514b/src/guest_memory.rs#L51).

I would define the error as `GuestMemoryError(GuestMemoryError)`.

Reply:
That is right. I will update this.